src/client/tls.rs

```diff
@@ -10,45 +10,32 @@ use futures::ready;
 use std::cmp;
 use std::{
     io,
-    mem::MaybeUninit,
     pin::Pin,
     task::{self, Poll},
 };
-use tokio::{
-    io::{AsyncRead, AsyncWrite},
-    net::TcpStream,
-};
 #[cfg(feature = "tls")]
-use tokio_tls::TlsStream;
+use async_native_tls::TlsStream;
 #[cfg(feature = "tls")]
 use tracing::{event, Level};
 
 /// A wrapper to handle either TLS or bare connections.
-pub(crate) enum MaybeTlsStream {
-    Raw(TcpStream),
+pub(crate) enum MaybeTlsStream<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> {
+    Raw(S),
     #[cfg(feature = "tls")]
-    Tls(TlsStream<TlsPreloginWrapper<TcpStream>>),
+    Tls(TlsStream<TlsPreloginWrapper<S>>),
 }
 
-impl MaybeTlsStream {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> MaybeTlsStream<S> {
     #[cfg(feature = "tls")]
-    pub fn into_inner(self) -> TcpStream {
+    pub fn into_inner(self) -> S {
         match self {
             Self::Raw(s) => s,
             Self::Tls(mut tls) => tls.get_mut().stream.take().unwrap(),
         }
     }
 }
 
-impl AsyncRead for MaybeTlsStream {
-    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [MaybeUninit<u8>]) -> bool {
-        match self {
-            MaybeTlsStream::Raw(s) => s.prepare_uninitialized_buffer(buf),
-            #[cfg(feature = "tls")]
-            MaybeTlsStream::Tls(s) => s.prepare_uninitialized_buffer(buf),
-        }
-    }
-
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncRead for MaybeTlsStream<S> {
```

Comment:
You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.